---
title: "RJB_DESeq2"
author: "Rachel Boyd, Paul Hook, Nelson Barrientos"
date: "2022-10-27"
output:
  pdf_document: default
  html_document: default
---
Load Packages:
```{r, include=FALSE}
library(DESeq2)
library(dplyr)
library(readr)
library(sva)
library(tidyverse)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
library(stringr)
library(data.table)
library(ggforce)
library(Rsubread)
library(pheatmap)
library(gplots)
library(tibble)
library(ggplot2)
```


Loading Data:
```{r}
#set working directory:
setwd("C:/Users/rboyd25/OneDrive - Johns Hopkins/Documents/Hopkins/McCallion/Papers/SN4741/Bulk_RNAseq")

#read count matrix:
mat <- read_tsv("gencode.vM24.counts", skip = 1)

#read count summary stats:
mat.summary <- read_tsv("gencode.vM24.counts.summary")

#read annotation data:
gene.meta <- read_tsv("gencode_vM24_annotation.txt",col_names = c("Geneid","symbol","MGI"))
gene.meta$MGI[is.na(gene.meta$MGI)] <- "." #Adds '.' for NA values

#Remove Chr, Start, End, Strand from mat and relabel replicates: 
mat.filter <- mat %>%
  dplyr::select(-Chr,-Start,-End,-Strand)
names(mat.filter)[3:10] <- c("37.1","37.2","37.3","37.4","39.1","39.2","39.3","39.4") #Adds colnames to mat.filter

#Merge mat.filter and gene.meta by 'Geneid':
mat.merge <- merge(mat.filter,gene.meta,by="Geneid")

#Rearrange mat.merge to Geneid, symbol, MGI, Length, replicate counts:
mat.final <- mat.merge %>%
  dplyr::select(Geneid,symbol,MGI,Length,everything())
```


Formatting:
```{r}
#Add 'anno' to mat.final
mat.final$anno <- paste0(mat.final$Geneid,"-",mat.final$symbol)

#Generate counts matrix with anno in first column and replicate counts in rest of columns
cts <- mat.final[,c(13,5:12)] 

#Change rownames of cts to anno
rownames(cts) <- cts$anno 

#Create final matrix with 'anno' as rownames and replicate counts as columns
cts.final <- as.matrix(cts[,-1]) 

#Create df of sample, temp, replicate information
test <- data.frame(sample=names(cts)[-1],
                   temp=factor(rep(c("37","39"),each=4)),
                   replicate=factor(1:4,1:4))

#Make sample the rownames for df
rownames(test) <- test[,1]

#Remove sample column to end up with samples as rownames and 2 columns: temp, replicate
coldata <- test[,-1] 

#Check to see that rownames of coldata are the same as colnames of cts.final
all(rownames(coldata) == colnames(cts.final)) 

# Generate dds (DESeqDataSet) of count data:
dds <- DESeqDataSetFromMatrix(countData = cts.final,
                              colData = coldata,
                              design = ~ temp)
# Estimate size factors
dds <- estimateSizeFactors(dds)
dds

saveRDS(dds,file = "dds.Rds")
```


**START HERE WITH RDS**
```{r}
dds <- readRDS("dds.Rds")
```

Plot PCA: variance stabilizing transformation of dds
```{r}
#PCA plot function
PCA_plot = function(df) {
  ggplot(df, aes(PC1, PC2, color=temp)) +
  geom_point(size=3) +
  scale_color_manual(values = c("#0F5E60", "#CE001B")) +
  xlab(paste0("Principal Component #1 [",percentVar[1],"%]")) +
  ylab(paste0("Principal Component #2 [",percentVar[2],"%]")) + 
  geom_text_repel(data = df, aes(label=paste0(temp, "-",replicate)), min.segment.length = Inf, color="black") +
  coord_fixed(ylim = c(-16.5,16.5), xlim = c(-16.5,16.5)) +
  theme_bw() +
  theme(
    legend.position = "none",
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_text(size = 10, color = "black"),
    axis.text.y = element_text(size = 10, color = "black"),
    panel.border = element_rect(color = "black"),
    aspect.ratio = 1
  )
}


vsd <- vst(dds, blind=FALSE)
pcaData.vst <- plotPCA(vsd, intgroup=c("temp", "replicate"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData.vst, "percentVar"))
pca.vst <- PCA_plot(pcaData.vst)

ggsave('bulkRNA_PCA.pdf', width = 7, height = 6, pca.vst, bg='transparent')
```

Plot Heatmaps:
```{r}
# Count matrix
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:100]
df <- as.data.frame(colData(dds)[,c("temp", "replicate")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)


#Sample-to-sample distances
jpeg("Bulk_distplot.jpg")
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- c("37-1","37-2","37-3","37-4","39-1","39-2","39-3","39-4")
rownames(sampleDistMatrix) <- paste(vsd$temp, vsd$replicate, sep="-")
colours = colorRampPalette( rev(brewer.pal(9,"Blues")) )(256)
heatmap.2(sampleDistMatrix, 
          trace="none", 
          col=colours, 
          sepwidth=c(0,0),
          sepcolor="black",
          colsep=0:9,
          rowsep=0:9)
dev.off()
```


Correlation Heatmap:
```{r}
# Format Data
counts <- counts(dds)
cormatrix <- round(cor(counts, method = "spearman"),2)
head(cormatrix)

library(reshape2)
melted_cormatrix <- melt(cormatrix)
head(melted_cormatrix)

# Get lower triangle of the correlation matrix
  get_lower_tri<-function(cormatrix){
    cormatrix[upper.tri(cormatrix)] <- NA
    return(cormatrix)
  }

# Get upper triangle of the correlation matrix
  get_upper_tri <- function(cormatrix){
    cormatrix[lower.tri(cormatrix)]<- NA
    return(cormatrix)
  } 
  upper_tri <- get_upper_tri(cormatrix)

melted_cormatrix <- melt(upper_tri, na.rm = TRUE)
melted_cormatrix$Var1<-as.character(melted_cormatrix$Var1)
melted_cormatrix$Var2<-as.character(melted_cormatrix$Var2)

ggheatmap <- ggplot(melted_cormatrix, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "#ecca00", high = "#ec2400", mid = "#ec5300",
 midpoint = 0.95, limit = c(0.9,1), space = "Lab",
 name="Pearson\nCorrelation") +
  theme_minimal()+
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1),axis.text.y = element_text(vjust = 1, 
    size = 12, hjust = 1))+
 coord_fixed()

Bulk_corrplot <- ggheatmap + 
geom_text(aes(Var2, Var1, label = value), color = "black", size = 4) +
theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.grid.major = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA, size = 1),
  panel.background = element_blank(),
  axis.ticks = element_blank(),
  legend.justification = c(1, 0),
  legend.position = c(0.4, 0.5))

ggsave('bulkRNA_Corrplot.pdf', width = 7, height = 6, Bulk_corrplot, bg='transparent')
```

Pre-Filtering:
```{r}
#Filter dds to include only rows with an average of at least 1 read for each sample
dds.data.filt <- dds[rowSums(counts(dds))/8 > 1,] 
dds.data.filt 

#Makes 37 the reference temp
dds.data.filt$temp<- relevel(dds.data.filt$temp, ref = "37") 
dds.data.filt2$temp<- relevel(dds.data.filt2$temp, ref = "37") 
```


Differential Expression Analysis (operates on raw counts):
1. Estimates size factors = reference sample is created that represents the counts on a typical sample in the dataset (using geometric mean). Ratios of sample to “psuedo reference” are calculated for each gene (assuming that most genes are not changing dramatically, this ratio should be ~1). Size factors are the median of these ratios for each sample (median is used so any outlier genes will not affect the normalization).
2. Corrects for library size using 'avgTxLength' from assays(dds)
3. Estimates dispersion (variability of each gene), gene-wise dispersion, mean-dispersion relationship, and final dispersion estimates
4. Model fitting and testing (data are count-based and not normally-distributed, so negative-binomial distribution is used)
```{r}
# DEseq with untransformed data:
dds.data.filt <- DESeq(dds.data.filt)

```


Filtering Results: "Results" function automatically performs independent filtering based on the mean of normalized counts for each gene, optimizing the number of genes which will have an adjusted p value below a given FDR cutoff, alpha.
```{r}
#Set the FDR cut-off to 0.01 (default 0.1)
res.plain <- results(dds.data.filt, alpha = 0.01)
# Re-order
res.plain <- res.plain[order(res.plain$padj),]
summary(res.plain)
res.plain
```


Log fold change shrinkage for visualization and ranking: "Tries to address the problem that lowly expressed genes tend to have high relatively levels of variability (it looks at the largest fold changes that are not due to low counts and uses these to inform a prior distribution. So the large fold changes from genes with lots of statistical information are not shrunk, while the imprecise fold changes are.). One case where I would not use it, is if it is expected that nearly all genes will have no change and there is little to no variation across replicates (so near technical replication), and then say < 10 genes with very large fold changes" [Michael Love]
```{r}
#Set the FDR cut-off to 0.01 (default 0.1)
res <- results(dds.data.filt, alpha = 0.01)

resultsNames(dds.data.filt)

#Pass results to lfc shrink
res <- lfcShrink(dds.data.filt, coef="temp_39_vs_37", type="apeglm", res=res)
# Re-order
res <- res[order(res$padj),]
head(res, n=10)
summary(res)
res
write.table(res, "DESeq2LFCshrinkage_results_FDR.0.01", quote = F)

res <- res[order(res$log2FoldChange),]
tail(res,n=20)
```

P-value distribution:
```{r}
hist.resLFC <- data_frame(pvalue = res$pvalue)
hist.plot <- ggplot(hist.resLFC, aes(x=pvalue)) +
        geom_histogram(binwidth = 0.02, fill = "Red", color = "Black") +
        xlab(paste0("P-Value")) +
        ylab(paste0("Frequency")) +
        theme_classic(base_size = 15)
ggsave('pval.dist.pdf', width = 7, height = 6,hist.plot, bg='transparent')
```

MA plots:
```{r}
# plot fold changes with and without shrinkage:
pdf("Bulk_MA.plots.pdf")
par(mfrow=c(2,1))
plotMA(res.plain, ylim=c(-8,8), ylab = ~log[2]~"Fold Change", xlab = "Mean of Normalized Counts")
plotMA(res, ylim=c(-8,8),ylab = ~log[2]~"Fold Change", xlab = "Mean of Normalized Counts")
dev.off()
```

Formatting for plotting:
```{r}
res.df <- as.data.frame(res)
res.df$isSig <- ifelse(between(res.df$log2FoldChange, -1.5, 1.5) | res.df$padj > 0.01, F, T)
res.df$genes <- str_split_fixed(rownames(res), "-", n = 2)[,2]
```

Volcano Plot
```{r}
# Volcano Plot Function
Volcano_plot = function(df, markers) {
  ggplot(df, aes(log2FoldChange, -log10(padj), color=isSig)) +
        geom_point(size=0.5, alpha=0.5) +
        scale_color_manual(values = c("#778187", "#CE001B")) +
        xlab(~log[2]~"Fold Change") +
        ylab(~-log[10]~"Adjusted P-Value") +
        scale_y_continuous(limits = c(-10, 310)) +
        scale_x_continuous(limits = c(-5, 5)) +
        geom_point(data = df[df$genes %in% markers$genes,],fill = "blue", shape = 21, size=2.5) + 
        geom_label_repel(data = df[df$gene %in% markers$genes,], aes(label=genes), color = "black",
                          size=5, fontface = "bold.italic", max.overlaps = Inf, 
                          min.segment.length = 0) +
        geom_vline(xintercept = c(-1.5,1.5), color="black", size=0.5, linetype=2) +
        geom_hline(yintercept = 2, color="black", size=0.5, linetype=2) +
        theme_bw() +
        theme(
          legend.position = "none",
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank(),
          panel.grid.major.y = element_blank(),
          panel.grid.minor.y = element_blank(),
          axis.text.x = element_text(size = 10, color = "black"),
          axis.title.x = element_text(size=15),
          axis.title.y = element_text(size=15),
          axis.text.y = element_text(size = 10, color = "black"),
          panel.border = element_rect(color = "black"))
}

# Comparison with Enhanced Volcano for Sarah:
blank <- data.frame(genes =c("abc"))
Volcano_plot(res.df, blank)

library(EnhancedVolcano)
enhanced <- EnhancedVolcano(res.df,
    lab = NA,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.01,
    FCcutoff = 1.5)
ggsave('enhanced_volcano.pdf', width = 10, height = 6, enhanced, bg='transparent')

#Volcano Plots:

# Immature neuron markers
IMMneurons <- data.frame(genes=c("Ntrk1","Unc13a","Cadm2","Nrxn1","Rax","Cacna1b","Celf5","Tmem179","Cux2","Dpysl5","Adgrb3","Nova2","Cntn2","Scn3b","Gpm6b","Sox5"))
ImmatureNeuronMarkers_Volcano <- Volcano_plot(res.df, IMMneurons)
ggsave('ImmNeuronMarkers_volcano.pdf', width = 10, height = 6, ImmatureNeuronMarkers_Volcano, bg='transparent')

#Pluripotent Stem Cell Markers
Pluri.cells <- data.frame(genes=c("Pif1","Suv39h2","Ercc6l","Bub1b","Mcm10","Kif11","Spn","Kntc1","Fbxo5","Bub1","Plk4","Fanci","Gins1","Hells","Cdc6","Ndc80","Ticrr","Sgo1","Ccna2","Depdc1b","Cenpf","Nasp","Fancd2","Aunip","Dna2","Mad2l1"))
PluricellMarkers_Volcano <- Volcano_plot(res.df, Pluri.cells)
ggsave('Pluri.cell.Markers_volcano.pdf', width = 10, height = 6, PluricellMarkers_Volcano, bg='transparent')


#Dopaminergic neuron markers
DAgenes <- data.frame(genes =c("Aldh1a1", "Foxa2", "Lmx1b", "Nr4a2", "Pitx3", "Slc6a3","Th", "Tmem179","Unc13a", "Dpysl5","Chrna4","Nrxn1","Nova2","Cacna1b","Cntn2","Celf5","Srcin1"))

DA.Volcano <- Volcano_plot(res.df, DAgenes)
ggsave('DAmarkers_volcano.pdf', width = 10, height = 6, DA.Volcano, bg='transparent')

#Oligodendrocyte markers
OLIGgenes <- data.frame(genes=c("Ptprt","Il33","Olig3","Arrdc2","Nrxn1","Hdac11","Dusp15","Srcin1","Nipal4","Efnb3","Psd2","Sec14l5","Enpp2","Cntn2","Sp7","Ptgds","Gpm6b"))
Olig.Volcano <- Volcano_plot(res.df, OLIGgenes)
ggsave('Olig.markers_volcano.pdf', width = 10, height = 6, Olig.Volcano, bg='transparent')

#Adrenergic neuron markers
ADRENgenes <- data.frame(genes=c("Pnmt", "Ddc", "Dbh", "Slc18a2","Npff","Slc12a7","Syt1","Th"))
Volcano_plot(res.df, ADRENgenes)

#Glutamatergic neuron markers
GLUTgenes <- data.frame(genes=c("Slc17a7","Slc17a6","Grin1","Grin2b","Gls","Glul"))
Volcano_plot(res.df, GLUTgenes)

#GABAergic neruon markers
GABAgenes <- data.frame(genes=c("Slc6a1", "Gabbr1", "Gabbr2", "Gad2", "Gad1"))
Volcano_plot(res.df, GABAgenes)

#Cholinergic neuron markers
CHOLINgenes <- data.frame(genes=c("Chat", "Slc18a3", "Ache"))
Volcano_plot(res.df, CHOLINgenes)

#Astrocyte markers
ASTROgenes <- data.frame(genes=c("Gfap", "Aldh1l1", "Glul","Slc1a3","Slc1a2","S100b"))
Volcano_plot(res.df, ASTROgenes)

#Highly expressed genes or transcripts 
of.interest <- data.frame(genes=c("Ooep","Gtf2a1l","Gm13284","Tmem132c","Krt8-ps","Klk6","Gm50056","Serpina3n","Otog","Ighe","H4c8","Eef1a2"))
Volcano_plot(res.df, of.interest)
```

Mouse E15.5 markers from Hook et al.,(2018)
```{r}
E15.FB.1 <- data.frame(genes=c("Dlx2","Dlx1","Dlx5","Fam84a","Gria1","Gabra5","Igfbpl1","Rnd3","Fgf14","Pnoc"))
Volcano_plot(res.df, E15.FB.1)

E15.FB.2 <- data.frame(genes=c("Gm5532","Npy","Six3os1","Gas1","Six3","Eepd1","Map3k1","Esrrg","Rpp25","Prox1","Fign","Sp8","Gm29374","Isl1","Sst","A830082K12Rik","Ecel1","Zeb2","Nr2f1","Slc44a5","Nr2f2","Gpc3","2900055J20Rik","Kcnmb2","Tmem132b","Zbtb7c","Unc5c","Nefm","Lrrtm1","Pde2a","Tenm3","Zfp804a","Galnt13","Hs3st5","Anks1","1110015O18Rik","Ubash3b","Tle1","Sema3a","Tshz2","Klhl5","Kcnma1","Zdhhc2","Gad2","Meis2","Klf12","Gad1","AI593442","Lrrc4c","Gm14204","Islr2","Tmem2","Id4","Cygb","Lhfp","Grik2","Rnf220","Rftn1","Vcan"))
Volcano_plot(res.df, E15.FB.2)

E15.MB.1 <- data.frame(genes=c("Pax5","Lhx9","Ebf1","Npnt","Tspan9","Nrn1","Nrg1","Gabra1","Sorcs2","Slc17a6","Mab21l1","Pcdh18","Syn2","Car10","Ppp1r14c","Fam19a2","Tmem163","Meis1","Rasgrp2","Cacna2d1","Pde4b","Lrrtm3","Wscd1","Surf6","Adap1","Mir99ahg","Pcdh17","Tmem91","Garnl3","Ckmt1"))
Volcano_plot(res.df, E15.MB.1)

E15.MB.2 <- data.frame(genes=c("Foxa1","Lmx1a","C130021I20Rik","Slc6a3","Chrna6","Pitx3","Foxa2","Gch1","Ntn1","En1","Schip1","Nr4a2","eGFP","Ret","Ddc","Slc18a2","Necab1","Wnt5a","Ctnna1","Th","Cpne7","Cnpy1","Syt17","Ptpru","Kcnip4","Uncx","Dlk1","Ebf2","Slc10a4","Foxp2","Adarb2","Klhl1","Serpine2","Pdzrn4","Igfbp5","Bnc2","Wdr54","St6galnac4","Tox2","Snca","Gm2694","Calb2","Tcf12","Rasgef1b","2810025M15Rik","Slc1a2","Cdk14","Sdc2","Pbx1","Syt13","Mt3","Ifitm2","Rit2","Mpped2","Slit2","Epha5","Epdr1","Ptprg","Osbpl5","Stk19"))
Volcano_plot(res.df, E15.MB.2)
```

GO of up-regulated and down-regulated genes:
```{r}
#Get lists of the genes up and down
downreg.genes <- subset(res, padj < 0.01 & log2FoldChange < -1.5)
upreg.genes <- subset(res, padj < 0.01 & log2FoldChange > 1.5)

#List of downregulated genes for enrichr
downreg.genes.GO <- str_split_fixed(row.names(downreg.genes), "-", 2)[,2] 
write.table(downreg.genes.GO, "Downregulated_genes_for_GO_1.5FC.txt", quote = F, row.names = F, col.names = F)

#List of upregulated genes for enrichr
upreg.genes.GO <- str_split_fixed(row.names(upreg.genes), "-", 2)[,2] 
write.table(upreg.genes.GO, "Upregulated_genes_for_GO_1.5FC.txt", quote = F, row.names = F, col.names = F)
```

GO Biological process enrichment & cell type plots:
```{r, eval=FALSE}
library(grid)
library(tidyverse)
library(shadowtext)
library(dplyr)

#Format Data for Plotting
GO_cell <- fread("for_GO.cells_graph.txt", header = T)
GO_cell <- as_tibble(GO_cell)
down_cell <- GO_cell %>% filter(facet=="cellDOWN")
up_cell <- GO_cell %>% filter(facet=="cellUP")
up_GO <- GO_cell %>% filter(facet=="upGO")
down_GO <- GO_cell %>% filter(facet=="downGO")

# Cell Type Markers (Downregulated)
downc<- down_cell %>% mutate(Term = fct_reorder(Term, Score)) %>% 
ggplot(aes(x=Term,y=Score, fill=bin_pval)) +
  geom_col(width=0.7) +
  scale_y_continuous(limits = c(0, 165)) +
  scale_fill_manual(values=c("#332288",
"#117733",
"#88ccee",
"#ddcc77",
"#cc6677",
"#aa4499")) +
  coord_flip() +
  xlab("") + 
  ylab("Combined Score Ranking") +
  labs(fill = "p-value", title="Downregulated PanglaoDB Cell Type Markers [Marker Gene Overlap]") +
  geom_vline(xintercept = 5.5, color="black", size=0.5, linetype=2) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0),
        legend.position = "none",
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 12))  +
  annotate("text", x = 6:5, y = 140, label = c("Top 10 Terms", "Terms of Interest"))
      
ggsave('Cell_downreg_bar.pdf', width = 10, height = 8, downc, bg='transparent')

# Cell Type Markers (Upregulated)
upc <- up_cell %>% mutate(Term = fct_reorder(Term, Score)) %>% 
  ggplot(aes(x=Term,y=Score, fill=bin_pval)) +
  geom_col(width=0.7) +
  scale_y_continuous(limits = c(0, 38)) +
  scale_fill_manual(values=c("#332288",
"#117733",
"#88ccee",
"#ddcc77",
"#cc6677",
"#aa4499")) +
  coord_flip() +
  xlab("") + 
  ylab("Combined Score Ranking") +
  labs(fill = "p-value",title="Upregulated PanglaoDB Cell Type Markers [Marker Gene Overlap]") +
  geom_vline(xintercept = 5.5, color="black", size=0.5, linetype=2) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0),
        legend.position = "none",
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 12))  +
  annotate("text", x = 6:5, y = 30, label = c("Top 10 Terms", "Terms of Interest"))
      
ggsave('Cell_upreg_bar.pdf', width = 10, height = 8, upc, bg='transparent')

# GO Bio (DownregulateD)
downgo <- down_GO %>% mutate(Term = fct_reorder(Term, Score)) %>% 
  ggplot(aes(x=Term,y=Score, fill=bin_pval)) +
  geom_col(width=0.7) +
  scale_y_continuous(limits = c(0, 660)) +
  scale_fill_manual(values=c("#332288",
"#117733",
"#88ccee",
"#ddcc77",
"#cc6677",
"#aa4499")) +
  coord_flip() +
  xlab("") + 
  ylab("Combined Score Ranking") +
  labs(fill = "p-value", title="Downregulated GO Biolgical Process [Marker Gene Overlap]") +
  geom_vline(xintercept = 5.5, color="black", size=0.5, linetype=2) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0),
        legend.position = "none",
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 12)) +
  annotate("text", x = 6:5, y = 630, label = c("Top 10 Terms", "Terms of Interest"))

      
ggsave('GO_downreg_bar.pdf', width = 10, height = 8, downgo, bg='transparent')

# GO Bio (UpregulateD)
upgo <- up_GO %>% mutate(Term = fct_reorder(Term, Score)) %>% 
  ggplot(aes(x=Term,y=Score, fill=bin_pval)) +
  geom_col(width=0.7) +
  scale_y_continuous(limits = c(0, 260)) +
  scale_fill_manual(values=c("#117733",
"#88ccee",
"#cc6677",
"#aa4499")) +
  coord_flip() +
  xlab("") + 
  ylab("Combined Score Ranking") +
  labs(fill = "p-value",title="Upregulated GO Biolgical Process [Marker Gene Overlap]") +
  geom_vline(xintercept = 7.5, color="black", size=0.5, linetype=2) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0),
        legend.position = "none",
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        axis.title.x = element_text(size = 12)) +
  annotate("text", x = 8:7, y = 200, label = c("Top 10 Terms", "Terms of Interest"))

ggsave('GO_upreg_bar.pdf', width = 10, height = 8, upgo, bg='transparent')
```

Converting Counts to RPKM for Comparison with E15.5 MB/FB RPKM (McClymont et al., 2018):
```{r}
# Calculate per million scaling factor:
scale37.1 <-sum(mat.final$`37.1`)/1000000
scale37.2 <-sum(mat.final$`37.2`)/1000000
scale37.3 <-sum(mat.final$`37.3`)/1000000
scale37.4 <-sum(mat.final$`37.4`)/1000000
scale39.1 <-sum(mat.final$`39.1`)/1000000
scale39.2 <-sum(mat.final$`39.2`)/1000000
scale39.3 <-sum(mat.final$`39.3`)/1000000
scale39.4 <-sum(mat.final$`39.4`)/1000000

#Divide the read counts by the “per million” scaling factor. This normalizes for sequencing depth, giving you reads per million (RPM)
mat.rpkm <- mat.final
mat.rpkm$`37.1` <- (mat.final$`37.1`/scale37.1)
mat.rpkm$`37.2` <- (mat.final$`37.2`/scale37.2)
mat.rpkm$`37.3` <- (mat.final$`37.3`/scale37.3)
mat.rpkm$`37.4` <- (mat.final$`37.4`/scale37.4)
mat.rpkm$`39.1` <- (mat.final$`39.1`/scale39.1)
mat.rpkm$`39.2` <- (mat.final$`39.2`/scale39.2)
mat.rpkm$`39.3` <- (mat.final$`39.3`/scale39.3)
mat.rpkm$`39.4` <- (mat.final$`39.4`/scale39.4)

#Divide the RPM values by the length of the gene, in kilobases. This gives you RPKM.
mat.rpkm[,5:12] <- mat.rpkm[,5:12]/mat.rpkm$Length

#Average RPKM for each temperature
mat.rpkm2 <- mat.rpkm
mat.rpkm2$rpkm.mean.37 <- rowMeans(mat.rpkm[ , c(5:8)], na.rm=TRUE)
mat.rpkm2$rpkm.mean.39 <- rowMeans(mat.rpkm[ , c(9:12)], na.rm=TRUE)

# Format mat.rpkm for merging:
temp1 = str_split_fixed(mat.rpkm$Geneid, "\\.", 2)
mat.rpkm3 <- mat.rpkm2[,-2:-12]
mat.rpkm3$Geneid <- temp1[,1]
colnames(mat.rpkm3)[colnames(mat.rpkm3) == "Geneid"] = "GeneID"

#Feed in FB and MB RPKM:
MB.FB.rpkm <- read_tsv("E15.5_MB_FB_All.RPKM.txt")

# Merge:
RPKM.corr <- merge(x=mat.rpkm3,y=MB.FB.rpkm,by="GeneID",all=TRUE)
RPKM.corr[is.na(RPKM.corr)] <- 0 

# Generate Correlation HeatMap:
RPKMcormatrix <- round(cor(RPKM.corr[,2:5], method = "spearman"),2)
head(RPKMcormatrix)

melted_RPKMcormatrix <- melt(RPKMcormatrix)
head(melted_RPKMcormatrix)

# Get lower triangle of the correlation matrix
  get_lower_tri<-function(RPKMcormatrix){
    RPKMcormatrix[upper.tri(RPKMcormatrix)] <- NA
    return(RPKMcormatrix)
  }

# Get upper triangle of the correlation matrix
  get_upper_tri <- function(RPKMcormatrix){
    RPKMcormatrix[lower.tri(RPKMcormatrix)]<- NA
    return(RPKMcormatrix)
  } 
  RPKM.upper_tri <- get_upper_tri(RPKMcormatrix)

melted_RPKMcormatrix <- melt(RPKM.upper_tri, na.rm = TRUE)
melted_RPKMcormatrix$Var1<-as.character(melted_RPKMcormatrix$Var1)
melted_RPKMcormatrix$Var2<-as.character(melted_RPKMcormatrix$Var2)

RPKMheatmap <- ggplot(melted_RPKMcormatrix, aes(Var2, Var1, fill = value))+
 geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "yellow", 
   midpoint = 0.77, limit = c(0.56,1), space = "Lab", 
 name="Pearson\nCorrelation") +
  theme_minimal()+
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1),axis.text.y = element_text(vjust = 1, 
    size = 12, hjust = 1))+
 coord_fixed()

RPKM_corrplot <- RPKMheatmap + 
geom_text(aes(Var2, Var1, label = value), color = "black", size = 4) +
theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.grid.major = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA, size = 1),
  panel.background = element_blank(),
  axis.ticks = element_blank(),
  legend.justification = c(1, 0),
  legend.position = c(1.3, 0))

ggsave('MB.FB.SN4741.RNA_Corrplot.pdf', width = 7, height = 6, RPKM_corrplot, bg='transparent')
```

